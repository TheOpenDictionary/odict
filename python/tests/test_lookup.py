import pytest
import tempfile
import os

from pathlib import Path
from pyodict import Dictionary


@pytest.fixture(scope="module")
def dict1_path():
    current_file = Path(__file__).resolve()
    return str(current_file.parent.parent.parent / "examples" / "example1.xml")


@pytest.fixture(scope="module")
def dict2_path():
    current_file = Path(__file__).resolve()
    return str(current_file.parent.parent.parent / "examples" / "example2.xml")


@pytest.fixture(scope="module")
def dict1(dict1_path):
    return Dictionary.compile(dict1_path)


@pytest.fixture(scope="module")
def dict2(dict2_path):
    return Dictionary.compile(dict2_path)


def test_dictionary_path(dict1, dict1_path, dict2, dict2_path):
    assert dict1_path.replace(".xml", "") in dict1.path
    assert dict2_path.replace(".xml", "") in dict2.path


def test_lookup(dict1, snapshot):
    result = dict1.lookup("cat (cat)")
    assert result == snapshot


def test_lookup_no_split(dict1):
    # Test lookup without splitting
    result = dict1.lookup("catdog")
    assert len(result[0]) == 0


def test_lookup_with_split(dict1, snapshot):
    # Test lookup with splitting
    result = dict1.lookup("catdog", split=3)
    print(result)
    assert result == snapshot  # Use snapshot testing


def test_lexicon(dict1):
    # Test lexicon retrieval
    result = dict1.lexicon()
    assert result == ["cat", "dog", "poo", "ran", "run"]


def test_write_raw_xml():
    # Test writing raw XML and creating a dictionary
    xml_content = """
    <dictionary>
        <entry term="hello">
            <ety>
                <sense pos="v">
                    <definition value="hello world" />
                </sense>
            </ety>
        </entry>
        <entry term="world">
            <ety>
                <sense pos="v">
                    <definition value="hello world" />
                </sense>
            </ety>
        </entry>
    </dictionary>
    """

    with tempfile.NamedTemporaryFile(suffix=".odict", delete=False) as temp_file:
        temp_path = temp_file.name

    Dictionary.write(xml_content, temp_path)

    try:
        assert os.path.exists(temp_path)

        dict_instance = Dictionary(temp_path)
        assert len(dict_instance.lookup("hello")) == 1

    finally:
        # Clean up the temporary file
        os.unlink(temp_path)


def test_split_terms(dict1, snapshot):
    # Test term splitting
    result = dict1.split("catdog", threshold=2)
    assert result == snapshot  # Use snapshot testing


def test_index_and_search(dict1, snapshot):
    # Test indexing and searching
    dict1.index()
    results = dict1.search("run")
    assert results == snapshot  # Use snapshot testing
