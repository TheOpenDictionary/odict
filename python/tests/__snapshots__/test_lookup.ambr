# serializer version: 1
# name: test_index_and_search
  list([
    Entry { term: "run", see_also: None, etymologies: [Etymology { id: None, pronunciation: None, description: Some("Latin root"), senses: [("n", Sense { pos: "n", definitions: [Left(Definition { id: None, value: "(vertebrates) To move swiftly", examples: [Example { value: "The dog runs after the cat." }, Example { value: "The horse ran away." }], notes: [] }), Left(Definition { id: None, value: "Act or instance of hurrying (to or from a place) (not necessarily by foot); dash or errand, trip.", examples: [], notes: [] }), Left(Definition { id: None, value: "A pleasure trip.", examples: [], notes: [] }), Left(Definition { id: None, value: "Flight, instance or period of fleeing.", examples: [], notes: [] }), Left(Definition { id: None, value: "Migration (of fish).", examples: [], notes: [] }), Left(Definition { id: None, value: "A group of fish that migrate, or ascend a river for the purpose of spawning.", examples: [], notes: [] })] }), ("v", Sense { pos: "v", definitions: [Right(Group { id: None, description: "A number of verb senses", definitions: [Definition { id: None, value: "(vertebrates) To move swiftly (figuratively)", examples: [Example { value: "The dog runs after the cat." }, Example { value: "The horse ran away." }], notes: [] }, Definition { id: None, value: "(fluids) To flow.", examples: [], notes: [] }, Definition { id: None, value: "(nautical, of a vessel) To sail before the wind, in distinction from reaching or sailing close-hauled.", examples: [], notes: [] }, Definition { id: None, value: "(social) To carry out an activity.", examples: [], notes: [] }, Definition { id: None, value: "To extend or persist, statically or dynamically, through space or time.", examples: [], notes: [] }, Definition { id: None, value: "(transitive) To execute or carry out a plan, procedure or program.", examples: [], notes: [] }] })] })] }] },
  ])
# ---
# name: test_lookup
  list([
    list([
      Entry { term: "cat", see_also: None, etymologies: [Etymology { id: None, pronunciation: None, description: Some("Latin root"), senses: [("n", Sense { pos: "n", definitions: [Left(Definition { id: None, value: "a cat", examples: [Example { value: "There goes a cat!" }], notes: [Note { id: None, value: "Some definition note", examples: [Example { value: "Some example" }] }] })] })] }] },
    ]),
  ])
# ---
# name: test_lookup_with_split
  list([
    list([
      Entry { term: "cat", see_also: None, etymologies: [Etymology { id: None, pronunciation: None, description: Some("Latin root"), senses: [("n", Sense { pos: "n", definitions: [Left(Definition { id: None, value: "a cat", examples: [Example { value: "There goes a cat!" }], notes: [Note { id: None, value: "Some definition note", examples: [Example { value: "Some example" }] }] })] })] }] },
      Entry { term: "dog", see_also: None, etymologies: [Etymology { id: None, pronunciation: Some("dooooog"), description: Some("Latin root"), senses: [("un", Sense { pos: "un", definitions: [Left(Definition { id: None, value: "a dog", examples: [], notes: [] })] })] }] },
    ]),
  ])
# ---
# name: test_split_terms
  list([
    Entry { term: "cat", see_also: None, etymologies: [Etymology { id: None, pronunciation: None, description: Some("Latin root"), senses: [("n", Sense { pos: "n", definitions: [Left(Definition { id: None, value: "a cat", examples: [Example { value: "There goes a cat!" }], notes: [Note { id: None, value: "Some definition note", examples: [Example { value: "Some example" }] }] })] })] }] },
    Entry { term: "dog", see_also: None, etymologies: [Etymology { id: None, pronunciation: Some("dooooog"), description: Some("Latin root"), senses: [("un", Sense { pos: "un", definitions: [Left(Definition { id: None, value: "a dog", examples: [], notes: [] })] })] }] },
  ])
# ---
